File: Problems/0322. (M) Coin Change/Python3/solution.py
class Solution:
  def coinChange(self, coins: List[int], amount: int) -> int:
    # bottom-up dynamic programming

    # trivial case
    if not amount:
      return 0

    # value of dp[i] is min number of coins that adds up to i
    dp = [float('inf')] * (amount+1)
    dp[0] = 0

    for i in range(1, amount+1):
      for denom in coins:
        if i >= denom:
          dp[i] = min(dp[i], dp[i-denom]+1)

    return dp[-1] if dp[-1] != float('inf') else -1

File: Problems/0322. (M) Coin Change/Java/solution.java
class Solution {
  public int coinChange(int[] coins, int amount) {
    // bottom-up dynamic programming

    // trivial case
    if (amount == 0)
      return 0;

    // value of dp[i] is min number of coins that add up to i
    int[] dp = new int[amount+1];
    Arrays.fill(dp, amount+1);
    dp[0] = 0;

    for (int i = 1; i <= amount; i++) {
      // Java for-each syntax
      for (int denom : coins) {
        if (i >= denom) {
          dp[i] = Math.min(dp[i], dp[i-denom]+1);
        }
      }
    }
    
    return dp[amount] <= amount ? dp[amount] : -1;
  }
}
